conjure-runtime/src/service/node/limiter/ciad.rs

```diff
@@ -0,0 +1,297 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use crate::service::map_error::RawClientError;
+use crate::service::node::limiter::deficit_semaphore::{self, DeficitSemaphore};
+use crate::util::atomic_f64::AtomicF64;
+use conjure_error::Error;
+use futures::ready;
+use http::{Response, StatusCode};
+use pin_project::pin_project;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+const INITIAL_LIMIT: usize = 20;
+const BACKOFF_RATIO: f64 = 0.9;
+const MIN_LIMIT: usize = 1;
+const MAX_LIMIT: usize = 1_000_000;
+
+/// An asynchronous concurrency limiter. "Ciad" stands for "cautious increase, agressive decrease", which is a mouthful!
+///
+/// The limiter is tagged with a "behavior" which controls how a response from the server affects the limit.
+pub struct CiadConcurrencyLimiter<B> {
+    semaphore: Arc<DeficitSemaphore>,
+    limit: AtomicF64,
+    in_flight: AtomicUsize,
+    _p: PhantomData<B>,
+}
+
+impl<B> CiadConcurrencyLimiter<B> {
+    pub fn new() -> Arc<Self> {
+        Arc::new(CiadConcurrencyLimiter {
+            semaphore: DeficitSemaphore::new(INITIAL_LIMIT),
+            limit: AtomicF64::new(INITIAL_LIMIT as f64),
+            in_flight: AtomicUsize::new(0),
+            _p: PhantomData,
+        })
+    }
+
+    pub fn limit(&self) -> f64 {
+        self.limit.load(Ordering::SeqCst)
+    }
+
+    pub fn in_flight(&self) -> usize {
+        self.in_flight.load(Ordering::SeqCst)
+    }
+
+    pub fn acquire(self: Arc<Self>) -> Acquire<B> {
+        Acquire {
+            future: self.semaphore.clone().acquire(),
+            limiter: self,
+        }
+    }
+
+    fn update_limit<F>(&self, f: F)
+    where
+        F: Fn(f64) -> Option<f64>,
+    {
+        let mut old_limit = self.limit.load(Ordering::SeqCst);
+
+        loop {
+            let new_limit = match f(old_limit) {
+                Some(new_limit) => new_limit,
+                None => return,
+            };
+
+            match self.limit.compare_exchange(
+                old_limit,
+                new_limit,
+                Ordering::SeqCst,
+                Ordering::SeqCst,
+            ) {
+                Ok(_) => {
+                    let old_limit = old_limit as usize;
+                    let new_limit = new_limit as usize;
+
+                    #[allow(clippy::comparison_chain)]
```

Comment:
Do we dislike comparison chain in general or specifically here? If we dislike it specifically here, why is it bad here?

Reply:
It just seemed a bit less readable to me I think:

```rust
match new_limit.cmp(&old_limit) {
    ops::Ordering::Greater => self.semaphore.add_permits(new_limit - old_limit),
    ops::Ordering::Less => self.semaphore.remove_permits(old_limit - new_limit),
    ops::Ordering::Equal => {}
}
```

Reply:
OK, how about we always allow `clippy::comparison_chain` so we don't have to specifically allow it?

Reply:
My instinct is to keep it warning globally unless we end up seeing more cases where we don't like the recommendation.